---------------
---------------
# ***------------------------------------- PART 1 / 3 -------------------------------------***
---------------
---------------

In [1]:
import numpy as np
import pandas as pd
import matplotlib.cm as cm
import matplotlib.colors as colors
from sklearn.cluster import KMeans
import requests
import geocoder
import folium

Let's retrieve the data from Wikipedia and import them into a dataframe with read_html:

In [2]:
url = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
toronto_raw = pd.read_html(url)

What is the type of this object ?

In [3]:
print(type(toronto_raw))

<class 'list'>


What is its length ?

In [4]:
print(len(toronto_raw))

3


Let's examine each element and try to find if one of the may correspond to our need...

In [5]:
first, second, third = toronto_raw
first

Postal Code           Borough  \
0           M1A      Not assigned   
1           M2A      Not assigned   
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
..          ...               ...   
175         M5Z      Not assigned   
176         M6Z      Not assigned   
177         M7Z      Not assigned   
178         M8Z         Etobicoke   
179         M9Z      Not assigned   

                                         Neighbourhood  
0                                         Not assigned  
1                                         Not assigned  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
..                                                 ...  
175                                       Not assigned  
176                                       Not assigned  
177                                       Not assigned  
178  Mimico NW, The Queensway West, South of Bloor,...  
179                                       Not assigned  

[180 rows x 3 columns]

second and third appeared to be much less interesting than first, so let's use this one:

In [6]:
toronto = first

And let's del rows where 'Borough' = 'Not assigned'

In [7]:
toronto = toronto[toronto['Borough'] != "Not assigned"]

In [8]:
toronto.head()

Postal Code           Borough                                Neighbourhood
2         M3A        North York                                    Parkwoods
3         M4A        North York                             Victoria Village
4         M5A  Downtown Toronto                    Regent Park, Harbourfront
5         M6A        North York             Lawrence Manor, Lawrence Heights
6         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

Now reset the indexes:

In [9]:
toronto.reset_index(drop=True, inplace=True)
toronto.head(10)

Postal Code           Borough                                Neighbourhood
0         M3A        North York                                    Parkwoods
1         M4A        North York                             Victoria Village
2         M5A  Downtown Toronto                    Regent Park, Harbourfront
3         M6A        North York             Lawrence Manor, Lawrence Heights
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government
5         M9A         Etobicoke      Islington Avenue, Humber Valley Village
6         M1B       Scarborough                               Malvern, Rouge
7         M3B        North York                                    Don Mills
8         M4B         East York              Parkview Hill, Woodbine Gardens
9         M5B  Downtown Toronto                     Garden District, Ryerson

Eventually, let's look at its shape :

In [10]:
toronto.shape

(103, 3)

---------------
---------------
# ***------------------------------------- PART 2 / 3 -------------------------------------***
---------------
---------------

In [11]:
# !pip install geocoder

Note : as Google API did not return any result, I used arcgis instead.

In [12]:
lat_list = []
long_list = []
i = 1
for code in toronto['Postal Code']:
    g = geocoder.arcgis(f'{code}, Toronto, Ontario')
    lat_list.append(g.json['lat'])
    long_list.append(g.json['lng'])
    if i%10 == 0: print(f'{i}...', end = ' ', flush=True)
    i += 1
print('Done!')   

10... 20... 30... 40... 50... 60... 70... 80... 90... 100... Done!


In [13]:
toronto.loc[:,'Latitude'] = lat_list
toronto.loc[:,'Longitude'] = long_list
toronto

c:\users\sophi\appdata\local\programs\python\python38-32\lib\site-packages\pandas\core\indexing.py:1596: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self.obj[key] = _infer_fill_value(value)
c:\users\sophi\appdata\local\programs\python\python38-32\lib\site-packages\pandas\core\indexing.py:1745: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  isetter(ilocs[0], value)


Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood  Latitude  Longitude  
0                                            Parkwoods  43.75245  -79.32991  
1                                     Victoria Village  43.73057  -79.31306  
2                            Regent Park, Harbourfront  43.65512  -79.36264  
3                     Lawrence Manor, Lawrence Heights  43.72327  -79.45042  
4          Queen's Park, Ontario Provincial Government  43.66253  -79.39188  
..                                                 ...       ...        ...  
98       The Kingsway, Montgomery Road, Old Mill North  43.65319  -79.51113  
99                                Church and Wellesley  43.66659  -79.38133  
100  Business reply mail Processing Centre, South C...  43.64869  -79.38544  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  43.63278  -79.48945  
102  Mimico NW, The Queensway West, South of Bloor,...  43.62513  -79.52681  

[103 rows x 5 columns]

Coordinates seem to be slightly different... (e.g : M5A Downton Toronto is 43.65512, -79.36264 instead of 43.654250, -79.360636 which represent about 190m).
Maybe because I did not use the Google API ?
To avoid any problem I would rather use the csv file data but it seems not to be available (Error 501).

---------------
---------------
# ***------------------------------------- PART 3 / 3 -------------------------------------***
---------------
---------------

Let's select only Boroughs that contain the word "Toronto":

In [14]:
toronto = toronto[toronto['Borough'].str.contains('Toronto')]
toronto

Postal Code           Borough  \
2           M5A  Downtown Toronto   
4           M7A  Downtown Toronto   
9           M5B  Downtown Toronto   
15          M5C  Downtown Toronto   
19          M4E      East Toronto   
20          M5E  Downtown Toronto   
24          M5G  Downtown Toronto   
25          M6G  Downtown Toronto   
30          M5H  Downtown Toronto   
31          M6H      West Toronto   
36          M5J  Downtown Toronto   
37          M6J      West Toronto   
41          M4K      East Toronto   
42          M5K  Downtown Toronto   
43          M6K      West Toronto   
47          M4L      East Toronto   
48          M5L  Downtown Toronto   
54          M4M      East Toronto   
61          M4N   Central Toronto   
62          M5N   Central Toronto   
67          M4P   Central Toronto   
68          M5P   Central Toronto   
69          M6P      West Toronto   
73          M4R   Central Toronto   
74          M5R   Central Toronto   
75          M6R      West Toronto   
79          M4S   Central Toronto   
80          M5S  Downtown Toronto   
81          M6S      West Toronto   
83          M4T   Central Toronto   
84          M5T  Downtown Toronto   
86          M4V   Central Toronto   
87          M5V  Downtown Toronto   
91          M4W  Downtown Toronto   
92          M5W  Downtown Toronto   
96          M4X  Downtown Toronto   
97          M5X  Downtown Toronto   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   

                                         Neighbourhood  Latitude  Longitude  
2                            Regent Park, Harbourfront  43.65512  -79.36264  
4          Queen's Park, Ontario Provincial Government  43.66253  -79.39188  
9                             Garden District, Ryerson  43.65739  -79.37804  
15                                      St. James Town  43.65215  -79.37587  
19                                         The Beaches  43.67709  -79.29547  
20                                         Berczy Park  43.64536  -79.37306  
24                                  Central Bay Street  43.65609  -79.38493  
25                                            Christie  43.66869  -79.42071  
30                            Richmond, Adelaide, King  43.64970  -79.38258  
31                        Dufferin, Dovercourt Village  43.66505  -79.43891  
36   Harbourfront East, Union Station, Toronto Islands  43.64285  -79.38076  
37                            Little Portugal, Trinity  43.64848  -79.41774  
41                        The Danforth West, Riverdale  43.68375  -79.35512  
42            Toronto Dominion Centre, Design Exchange  43.64710  -79.38153  
43        Brockton, Parkdale Village, Exhibition Place  43.63941  -79.42676  
47                      India Bazaar, The Beaches West  43.66797  -79.31467  
48                      Commerce Court, Victoria Hotel  43.64840  -79.37914  
54                                     Studio District  43.66213  -79.33497  
61                                       Lawrence Park  43.72843  -79.38713  
62                                            Roselawn  43.71208  -79.41848  
67                                    Davisville North  43.71276  -79.38851  
68     Forest Hill North & West, Forest Hill Road Park  43.69479  -79.41440  
69                       High Park, The Junction South  43.65973  -79.46281  
73                   North Toronto West, Lawrence Park  43.71458  -79.40668  
74                 The Annex, North Midtown, Yorkville  43.67484  -79.40452  
75                              Parkdale, Roncesvalles  43.64777  -79.44989  
79                                          Davisville  43.70340  -79.38659  
80                      University of Toronto, Harbord  43.66311  -79.40180  
81                                  Runnymede, Swansea  43.64982  -79.47548  
83                         Moore Park, Summerhill East  43.69048  -79.38318  
84           Kensington Market, Chinatown, Grange Park  43.65351  -79.39722  
86   Summerhill West, Rathn

In [15]:
CLIENT_ID = 'H0LRD2VTB1XWM5F3RKUTSLLC54FLOXC1TAIPNMMTS5B1ITFO' # your Foursquare ID
CLIENT_SECRET = '1RXFUVNB0ZKVKOLGCVTMHOUJY2U4HQQMT2ZDHSHCSFML30QJ' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version
LIMIT = 100 # A default Foursquare API limit value

In [16]:
def get_nearby_venues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
           
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [17]:
toronto_venues = get_nearby_venues(names=toronto['Neighbourhood'],
                                  latitudes=toronto['Latitude'], 
                                  longitudes=toronto['Longitude']) 

In [18]:
print(toronto_venues.shape)

(1752, 7)


In [19]:
toronto_venues.head()

Neighbourhood  Neighborhood Latitude  Neighborhood Longitude  \
0  Regent Park, Harbourfront               43.65512               -79.36264   
1  Regent Park, Harbourfront               43.65512               -79.36264   
2  Regent Park, Harbourfront               43.65512               -79.36264   
3  Regent Park, Harbourfront               43.65512               -79.36264   
4  Regent Park, Harbourfront               43.65512               -79.36264   

                    Venue  Venue Latitude  Venue Longitude  Venue Category  
0        Roselle Desserts       43.653447       -79.362017          Bakery  
1           Tandem Coffee       43.653559       -79.361809     Coffee Shop  
2  Figs Breakfast & Lunch       43.655675       -79.364503  Breakfast Spot  
3         The Yoga Lounge       43.655515       -79.364955     Yoga Studio  
4     Body Blitz Spa East       43.654735       -79.359874             Spa

In [20]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighbourhood'] = toronto_venues['Neighbourhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Neighbourhood  Accessories Store  Afghan Restaurant  \
0  Regent Park, Harbourfront                  0                  0   
1  Regent Park, Harbourfront                  0                  0   
2  Regent Park, Harbourfront                  0                  0   
3  Regent Park, Harbourfront                  0                  0   
4  Regent Park, Harbourfront                  0                  0   

   American Restaurant  Antique Shop  Aquarium  Art Gallery  Art Museum  \
0                    0             0         0            0           0   
1                    0             0         0            0           0   
2                    0             0         0            0           0   
3                    0             0         0            0           0   
4                    0             0         0            0           0   

   Arts & Crafts Store  Asian Restaurant  ...  Train Station  \
0                    0                 0  ...              0   
1                    0                 0  ...              0   
2                    0                 0  ...              0   
3                    0                 0  ...              0   
4                    0                 0  ...              0   

   Vegetarian / Vegan Restaurant  Veterinarian  Video Game Store  \
0                              0             0                 0   
1                              0             0                 0   
2                              0             0                 0   
3                              0             0                 0   
4                              0             0                 0   

   Vietnamese Restaurant  Wine Bar  Wine Shop  Wings Joint  Women's Store  \
0                      0         0          0            0              0   
1                      0         0          0            0              0   
2                      0         0          0            0              0   
3                      0         0          0            0              0   
4                      0         0          0            0              0   

   Yoga Studio  
0            0  
1            0  
2            0  
3            1  
4            0  

[5 rows x 233 columns]

In [21]:
toronto_grouped = toronto_onehot.groupby('Neighbourhood').mean().reset_index()
toronto_grouped.head()

Neighbourhood  Accessories Store  \
0                                        Berczy Park           0.000000   
1       Brockton, Parkdale Village, Exhibition Place           0.011494   
2  Business reply mail Processing Centre, South C...           0.000000   
3  CN Tower, King and Spadina, Railway Lands, Har...           0.000000   
4                                 Central Bay Street           0.000000   

   Afghan Restaurant  American Restaurant  Antique Shop  Aquarium  \
0                0.0                 0.00      0.016667       0.0   
1                0.0                 0.00      0.000000       0.0   
2                0.0                 0.02      0.000000       0.0   
3                0.0                 0.00      0.000000       0.0   
4                0.0                 0.00      0.000000       0.0   

   Art Gallery  Art Museum  Arts & Crafts Store  Asian Restaurant  ...  \
0     0.016667    0.000000             0.000000          0.000000  ...   
1     0.000000    0.000000             0.022989          0.011494  ...   
2     0.010000    0.000000             0.010000          0.030000  ...   
3     0.000000    0.000000             0.000000          0.013158  ...   
4     0.013158    0.013158             0.000000          0.000000  ...   

   Train Station  Vegetarian / Vegan Restaurant  Veterinarian  \
0       0.000000                       0.016667      0.000000   
1       0.000000                       0.011494      0.000000   
2       0.000000                       0.020000      0.000000   
3       0.013158                       0.000000      0.013158   
4       0.000000                       0.000000      0.000000   

   Video Game Store  Vietnamese Restaurant  Wine Bar  Wine Shop  Wings Joint  \
0          0.000000               0.000000  0.000000        0.0          0.0   
1          0.000000               0.000000  0.000000        0.0          0.0   
2          0.000000               0.000000  0.010000        0.0          0.0   
3          0.000000               0.000000  0.000000        0.0          0.0   
4          0.013158               0.013158  0.013158        0.0          0.0   

   Women's Store  Yoga Studio  
0            0.0     0.016667  
1            0.0     0.011494  
2            0.0     0.000000  
3            0.0     0.013158  
4            0.0     0.000000  

[5 rows x 233 columns]

In [22]:
num_top_venues = 5

for hood in toronto_grouped['Neighbourhood']:
    temp = toronto_grouped[toronto_grouped['Neighbourhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})

In [23]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [24]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighbourhood'] = toronto_grouped['Neighbourhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()


Neighbourhood 1st Most Common Venue  \
0                                        Berczy Park           Coffee Shop   
1       Brockton, Parkdale Village, Exhibition Place                  Café   
2  Business reply mail Processing Centre, South C...           Coffee Shop   
3  CN Tower, King and Spadina, Railway Lands, Har...    Italian Restaurant   
4                                 Central Bay Street           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0            Restaurant                Bakery        Farmers Market   
1                   Bar           Coffee Shop            Restaurant   
2                 Hotel            Restaurant    Italian Restaurant   
3           Coffee Shop                  Café                   Bar   
4        Clothing Store                 Plaza             Bookstore   

  5th Most Common Venue      6th Most Common Venue 7th Most Common Venue  \
0          Cocktail Bar                   Beer Bar           Cheese Shop   
1        Sandwich Place                  Gift Shop             Nightclub   
2                  Café           Asian Restaurant                   Bar   
3                  Park          French Restaurant            Restaurant   
4                 Hotel  Middle Eastern Restaurant        Sandwich Place   

  8th Most Common Venue 9th Most Common Venue  10th Most Common Venue  
0        Breakfast Spot    Seafood Restaurant                     Pub  
1           Supermarket   Japanese Restaurant  Furniture / Home Store  
2      Sushi Restaurant    Seafood Restaurant                     Pub  
3             Speakeasy        Sandwich Place                  Bakery  
4                  Café            Restaurant        Sushi Restaurant

In [25]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
labels = kmeans.labels_ 
labels

array([4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 2, 4, 4, 1, 4, 4, 3, 4, 1, 1,
       4, 4, 4, 4, 1, 4, 4, 4, 4, 4, 0, 4, 1, 4, 4, 4])

In [26]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster', labels)
toronto_merged = toronto
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')
toronto_merged.dropna(inplace=True)
toronto_merged['Cluster']= toronto_merged['Cluster'].astype('int')

In [27]:
# create map
map_clusters = folium.Map(location=[43.651070, -79.347015], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'],toronto_merged['Neighbourhood'], toronto_merged['Cluster']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        #fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters